scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5159 @@
+        domain. Consequently, it may be preferred to work with the logarithms of
+        probabilities and probability densities to avoid underflow.
+
+        Examples
+        --------
+        Instantiate a distribution with the desired parameters:
+
+        >>> import numpy as np
+        >>> from scipy import stats
+        >>> X = stats.Uniform(a=-1.0, b=1.0)
+
+        Evaluate the log-PDF at the desired argument:
+
+        >>> X.logpdf(0.5)
+        -0.6931471805599453
+        >>> np.allclose(X.logpdf(0.5), np.log(X.pdf(0.5)))
+        True
+
+        """
+        return self._logpdf_dispatch(x, method=method, **self._parameters)
+
+    @_dispatch
+    def _logpdf_dispatch(self, x, *, method=None, **kwargs):
+        if self._overrides('_logpdf_formula'):
+            method = self._logpdf_formula
+        elif self.tol is _null:  # ensure that developers override _logpdf
+            method = self._logpdf_logexp
+        return method
+
+    def _logpdf_formula(self, x, **kwargs):
+        raise NotImplementedError(self._not_implemented)
+
+    def _logpdf_logexp(self, x, **kwargs):
+        return np.log(self._pdf_dispatch(x, **kwargs))
+
+    @_set_invalid_nan
+    def pdf(self, x, *, method=None):
+        r"""Probability density function
+
+        The probability density function :math:`f(x)` is the probability *per
+        unit interval* of :math:`x` that the random variable takes on the value
+        :math:`x`. Mathematically, it can be defined as the derivative of the
+        cumulative distribution function `F(x)`:
+
+        .. math::
+
+            f(x) = \frac{d}{dx} F(x)
+
+        `pdf` accepts `x` for :math:`x`.
+
+        Parameters
+        ----------
+        x : Array
+            The value of the random variable for which the PDF is to be
+            evaluated.
+        method : {None, 'formula', 'logexp'}
+            The strategy used to evaluate the PDF. By default (``None``), the
+            infrastructure chooses between the following options, listed in
+            order of precedence.
+
+            - ``'formula'``: use a formula for the PDF itself
+            - ``'logexp'``: evaluate the logarithm of the PDF directly and
+                            exponentiate
+
+            Not all `method` options are available for all distributions.
+            If the selected `method` is not available, a ``NotImplementedError``
+            will be raised.
+
+        Returns
+        -------
+        out : Array
+            The PDF evaluated at the argument `x`.
+
+        See Also
+        --------
+        cdf
+        logpdf
+
+        Notes
+        -----
+        Suppose continuous probability distribution has support :math:`[l, r]`.
+        By definition of the support, the PDF evaluates to its minimum value
+        of :math:`0` outside the support; i.e. for :math:`x < l` or
+        :math:`x > r`. The maximum of the PDF may be less than or greater than
```

Comment:
Would be good to clarify whether the support is necessarily a closes interval (endpoints). If it is, consider adding a means of making it flexible a feature request from me :-)

Reply:
I meant to check on that. There is some support, but I need to check the status.

Reply:
Yes, there is support.

Defining the `Uniform` distribution with:
```python3
    _x_support = _RealDomain(endpoints=('a', 'b'), inclusive=(True, False))
```
Then:
```python3
X = Uniform(a=0, b=1)
print(X._variable)  # `x` for :math:`x ∈ [a, b)`  (automatically appears in docstring)
X.pdf([0, 1])  # array([1., 0.])
X.logpdf([0, 1])  array([  0., -inf])
```
Is there anything else this should change?